ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HivePointLookupOptimizerRule.java

```diff
@@ -678,100 +679,135 @@ private RexNode useStructIfNeeded(List<? extends RexNode> columns) {
+          if (c == null || !HiveCalciteUtil.isDeterministic(c.exprNode)) {
+            continue;
+          }
+          if (c.exprNode.getType().isNullable()) {
+            inLHSExprToRHSNullableExprs.put(c.exprNode, c.exprNode);
+          }
+          if (c.constNode.getType().isNullable()) {
+            inLHSExprToRHSNullableExprs.put(c.exprNode, c.constNode);
+          }
+          if (inLHSExprToRHSExprs.containsKey(c.exprNode)) {
+            inLHSExprToRHSExprs.get(c.exprNode).retainAll(Collections.singleton(c.constNode));
+            if (!inLHSExprToRHSExprs.containsKey(c.exprNode)) {
+              // Note that Multimap does not keep a key if all its values are removed.
+              return createResultFromEmptySet(rexBuilder, c.exprNode, inLHSExprToRHSNullableExprs);
             }
+          } else {
+            inLHSExprToRHSExprs.put(c.exprNode, c.constNode);
           }
-          // Create IN clauses
-          newOperands = createInClauses(rexBuilder, inLHSExprToRHSExprs);
-          newOperands.addAll(operands);
-          // Return node
-          node = RexUtil.composeDisjunction(rexBuilder, newOperands, false);
-          break;
-        default:
-          return super.visitCall(call);
+          operands.remove(i);
+          --i;
+        }
       }
-      return node;
+      // Create IN clauses
+      final List<RexNode> newOperands = createInClauses(rexBuilder, inLHSExprToRHSExprs);
+      newOperands.addAll(operands);
+      // Return node
+      return RexUtil.composeConjunction(rexBuilder, newOperands, false);
+    }
+
+    private static RexNode handleOR(RexBuilder rexBuilder, RexCall call) {
+      // IN clauses need to be combined by keeping all elements
+      final List<RexNode> operands = new ArrayList<>(RexUtil.flattenOr(call.getOperands()));
+      final Multimap<RexNode,RexNode> inLHSExprToRHSExprs = LinkedHashMultimap.create();
+      for (int i = 0; i < operands.size(); i++) {
+        RexNode operand = operands.get(i);
+        if (operand.getKind() == SqlKind.IN) {
+          RexCall inCall = (RexCall) operand;
+          if (!HiveCalciteUtil.isDeterministic(inCall.getOperands().get(0))) {
+            continue;
+          }
+          RexNode ref = inCall.getOperands().get(0);
+          for (int j = 1; j < inCall.getOperands().size(); j++) {
+            inLHSExprToRHSExprs.put(ref, inCall.getOperands().get(j));
+          }
+          operands.remove(i);
+          --i;
+        }
+      }
+      // Create IN clauses
+      final List<RexNode> newOperands = createInClauses(rexBuilder, inLHSExprToRHSExprs);
+      newOperands.addAll(operands);
+      // Return node
+      return RexUtil.composeDisjunction(rexBuilder, newOperands, false);
+    }
+
+    private static RexNode createResultFromEmptySet(RexBuilder rexBuilder,
+        RexNode ref, Multimap<RexNode, RexNode> inLHSExprToRHSNullableExprs) {
+      if (inLHSExprToRHSNullableExprs.containsKey(ref)) {
```

Comment:
right now I don't fully agree with this logic - if we have an expression like:
```
x IN (1,2) AND x in (3,4) and y IN (5,6)
```
it's clear that we may only return `false` or `null`; but I think we may still need to choose `false` based on the other expressions; for the above:
* when `x is null, y=5` then the result will be `null`
* but when `x is null, y=1` then the result will be `false`

in light of this reasoning it seem to me that this starts looking more and more sophisticated....
I think it might worth a try to see if these "early-return-type" simplifications kick-in a lot or not - and as an alternative approach consider removing them instead of saving them.

I think at some point we should run the simplification earlier when the `OR`s are still open - so that it could analyze-it properly

Reply:
I agree. Isn't enough to move the logic to deal with the `empty` case to the end, instead of doing early bail out? I think that would fix the issue. I need to deal with the 'empty' case in some way.

I prefer to avoid moving the rule at this stage and keep the change focused within the rule. Otherwise, we may have to deal with closing and re-opening, plus the fix seems more risky (including backporting).

Reply:
> isn't enough to move the logic to deal with the empty case to the end

of course - that could work as well! I just wanted to suggest that in case there is no "big benefit" keeping it - we may remove it as well - it contained correctness issues before this patch :)

> I prefer to avoid moving the rule at this stage and keep the change focused within the rule.

I only wanted to note that we should do that sometime later on

Reply:
@kgyrtkirk I pushed a new commit and added a few more test cases.